File: 代码/KVO/03-KVO的Key依赖/01-KVO简介/ViewController.m
//
//  ViewController.m
//  01-KVO简介
//
//  Created by TT-Fangss on 2021/10/9.
//

#import "ViewController.h"
#import "Person.h"

@interface ViewController ()
@property (nonatomic, strong) Person *student;
@property (weak, nonatomic) IBOutlet UILabel *nameLabel;

@end

@implementation ViewController

- (Person *)student {
    if (!_student) {
        _student = [[Person alloc] init];
        _student.firstName = @"zhang";
        _student.lastName = @"shan";
    }
    return _student;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self registerAsObserver];
}

- (void)dealloc {
    [self unregisterAsObserver];
}

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
                       context:(void *)context {
    NSLog(@"\n-------\nkeyPath：%@，\nobject：%@，\nchange：%@，\ncontext：%@\n-------", keyPath, object, change, context);
}

- (IBAction)changedClick {
    if ([_student.fullName isEqualToString:@"zhang shan"]) {
        _student.firstName = @"li";
        _student.lastName = @"si";
    }else if ([_student.fullName isEqualToString:@"li si"]) {
        _student.firstName = @"wang";
        _student.lastName = @"wu";
    }else if ([_student.fullName isEqualToString:@"wang wu"]) {
        _student.firstName = @"zhang";
        _student.lastName = @"shan";
    }

    _nameLabel.text = _student.fullName;
}

- (IBAction)registerAsObserver {
    [self.student addObserver:self forKeyPath:@"fullName" options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld context:nil];
}

- (IBAction)unregisterAsObserver {
    @try {
        [self.student removeObserver:self forKeyPath:@"fullName"];
    } @catch (NSException *exception) {
        NSLog(@"\n------\nname：%@，\nreason：%@，\nuserInfo：%@------", exception.name, exception.reason, exception.userInfo);
    } @finally {
        NSLog(@"不管是否抛出异常，都会执行");
    }
}

@end

File: 代码/KVO/04-KVO原理-Swift/04-KVO原理-Swift/ViewController.swift
//
//  ViewController.swift
//  04-KVO原理-Swift
//
//  Created by TT-Fangss on 2021/10/12.
//

import UIKit

@objcMembers class Person: NSObject {
    @objc dynamic var age: NSInteger = 0
    
    func printObjectInfo() {
        print("-----")
        print("对象：\(self)")
        
        let cls_object: AnyClass = object_getClass(self)!  // 类对象
        let super_cls_object: AnyClass = class_getSuperclass(cls_object)!  // 类对象的父类对象（person->superclass_isa）
        let meta_cls_object: AnyClass = object_getClass(cls_object)!  // 元类对象（person->isa->isa）
        print("class 对象: \(cls_object)");
        print("class 对象的 superclass 对象: \(super_cls_object)");
        print("metaclass 对象: \(meta_cls_object)");

        let age_sel = #selector(setter: self.age)
        let age_imp = method(for: age_sel)
        print("setAge: \(String(describing: age_imp))")
        
        printMethodNamesOfClass(cls: cls_object)
    }
}

func printMethodNamesOfClass(cls: AnyClass) {
    var count: UInt32 = 0
    var methodNames = [String]()
    if let methodList = class_copyMethodList(cls, &count) {
        for i in 0..<count {
            let method = methodList[Int(i)]
            let sel = method_getName(method)
            let methodName = NSStringFromSelector(sel)
            methodNames.append(methodName)
        }
        
        free(methodList)
        
        print("对象方法列表: \(methodNames)")
    }
    
}

class ViewController: UIViewController {
    @IBOutlet weak var nameLabel: UILabel!
    @objc var person = Person()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
    }
    
    deinit {
        self.unregisterAsObserver()
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        print("keyPath: \(keyPath ?? ""), object: \(object ?? ""), change: \(String(describing: change))")
    }
    
    @IBAction func changedClick() {
        person.age = 10
        nameLabel.text = "\(person.age)"
    }
    
    @IBAction func registerAsObserver() {
        print("添加 Observer 之前")
        person.printObjectInfo()
        
        person.addObserver(self, forKeyPath: "age", options: [.new, .old], context: nil)
        
        print("添加 Observer 之后")
        person.printObjectInfo()
        person.age = 10
        print("结束")
    }
    
    @IBAction func unregisterAsObserver() {
        person.removeObserver(self, forKeyPath: "age")
        person.printObjectInfo()
        
//        let name: String = "_4_KVO原理_Swift.Person"
//        let cls: AnyClass = objc_getClass(name.cString(using: .utf8)!)! as! AnyClass
//        printMethodNamesOfClass(cls: cls)
        
        print("--")
    }
    
}

